wikimedia/wikipedia-ios
# Language variant alerts

**Fixes Phabricator ticket:** 
https://phabricator.wikimedia.org/T270210

### Notes
I think reviewing each commit individually should be a relatively straightforward review each step of the way.

Some key commits that bear further explanation:
**Commit 1 MWKDataStore changes:**
Since the same list of language codes is used for the variant data migration and the variant alerts, refactoring so that the list can be looked up by the library version where those languages became 'variant-aware'.

**Commit 3 NSUserActivity**
Added a new NSUserActivity to navigate to the app language settings. The navigation code is modeled on the settings view controller navigation method when the Languages settings is tapped.

**Commit 6 Present Variant Alerts**
This does the presentation of the alerts including chaining the alerts in the rare case a user has more than one language with a variant in their preferred languages.

The code seems to be called at the correct time in the app lifecycle, but if there is a better place to trigger this, I am very open to feedback.

### Test Steps
1. Set up a version of the shipping version of the app and ensure at least one language that will support variants is in the app's preferred language list. (Chinese, Serbian)
2. Upgrade to a build with this change the following should happen:
    - If a single language with variants is present in the app's preferred languages, one alert will be 
    displayed to the user. The alert should have "Review your preferences" button. Tapping dismisses alert and navigates to language preferences. "No thanks" button dismisses.

    - If more than one language with variants is displayed, multiple alerts, one per language, will be displayed sequentially. The initial alerts should have one "Got it" button which dismisses and presents the next alert. The final alert should have two buttons and behave as in the previous step.
3. After the migration, the alerts should not appear again on subsequent runs of the app.
